### PR TITLE
Rolling updater enhancements

### LINKED ISSUE #2828: Rolling-update should have a tunable for the % of pods to bring up at once
Make the rolling update scale amount configurable as a percent of the replica count.

Introduce a new optional deployment API field `strategy.rollingParams.updatePercent`. If specified, the amount of replicas scaled up and down each interval will be computed as a percentage of the desired replicas (for the new deployment) and the actual replicas (of the old deployment) respectively. If the field is nil, one replica will be scaled up and down each interval.

Depends on https://github.com/GoogleCloudPlatform/kubernetes/pull/10062.

### LINKED ISSUE #3313: Canary check is skipped during rolling update when "from" deployment has 0 replicas
Followup to #3308. The upstream rolling updater code will skip the canary check if the "from" deployment has zero replicas. We're carrying a patch for the canary code, which will need updated.

### PR DESCRIPTION
**Pull request by ironcladlou**
Integrate with upstream rolling update enhancements.
    
* Tweak the update acceptor to support being invoked on every update
      interval for a deployment.
* Update rolling params API to support updatePercent passed through
      to the upstream rolling updater.

Integrates deployments with https://github.com/GoogleCloudPlatform/kubernetes/pull/10062.

Closes #2828.
Fixes #3313.

### CODE REVIEW

**Comment by ironcladlou** [comment]
@kargakis @deads2k PTAL.

**Comment by deads2k** [comment]
Did you regenerate conversions too?

File: pkg/deploy/api/types.go [hunk 1]
```diff
@@ -115,6 +115,9 @@ type RollingDeploymentStrategyParams struct {
 	// TimeoutSeconds is the time to wait for updates before giving up. If the
 	// value is nil, a default will be used.
 	TimeoutSeconds *int64
+	// UpdatePercent is the percentage of replicas to scale up or down each
+	// interval. If nil, one replica will be scaled up and down each interval.
+	UpdatePercent *int `json:"updatePercent,omitempty" description:"the percentage of replicas to scale up or down each interval"`
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: This should have fail unit tests.  No json on internal types

Comment (ironcladlou) [inline pkg/deploy/api/types.go]: What unit test should have failed it? They're passing.

File: pkg/deploy/api/types.go [hunk 2]
```diff
@@ -115,6 +115,9 @@ type RollingDeploymentStrategyParams struct {
 	// TimeoutSeconds is the time to wait for updates before giving up. If the
 	// value is nil, a default will be used.
 	TimeoutSeconds *int64
+	// UpdatePercent is the percentage of replicas to scale up or down each
+	// interval. If nil, one replica will be scaled up and down each interval.
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: I probably missed other discussion, but is one the most reasonable default?  It seems a little weird to specify a percentage, but if you don't specify a percentage, we don't choose a default percentage, we choose an absolute value.

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/deploy/api/types.go]: Fixed

File: pkg/deploy/api/types.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/deploy/api/types.go]: The upstream code is hard-coded to scale by 1 up/down. I didn't want to change the existing default behavior to be a percent.

File: pkg/deploy/api/types.go (hunk 1 above)
Comment (deads2k) [inline pkg/deploy/api/types.go]: > What unit test should have failed it? They're passing.

https://github.com/openshift/origin/blob/master/pkg/api/latest/tags_test.go#L61

File: pkg/deploy/strategy/support/lifecycle.go [hunk 3]
```diff
@@ -262,11 +279,13 @@ func (c *FirstContainerReady) Accept(deployment *kapi.ReplicationController) err
...
-	// If we found multiple, use the first one and log a warning.
-	// TODO: should finding multiple be an error?
-	pod := &pods.Items[0]
-	if len(pods.Items) > 1 {
-		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
+	// Use the first unknown pod we find, and track all pods we see.
+	var pod *kapi.Pod
+	for _, candidate := range pods.Items {
+		if !c.knownPods.Has(candidate.Name) && pod == nil {
+			pod = &candidate
```
Comment (deads2k) [inline pkg/deploy/strategy/support/lifecycle.go]: Why don't we break here?  We're going to check only one pod, but claim we know about all of the pods?

Comment (ironcladlou) [inline pkg/deploy/strategy/support/lifecycle.go]: Yes: we need to keep track of all pods known for the controller so the next time we do verification we don't re-check a pod from the previous set.

**Comment by deads2k** [comment]
I wasn't involved the initial review of this item. This pull looks like minor fitting of an existing feature for upstream refactoring.  The fitting looks pretty reasonable, but I think you'll need to get to the original reviewer to sign off.  Could you also link me to the original pull?

**Comment by ironcladlou** [comment]
@kargakis can review this in more depth.

**Comment by ironcladlou** [comment]
@smarterclayton I'd also like you to review the UpdateAcceptor implementation change to make sure it aligns with the new semantics (to verify on every interval).

**Comment by smarterclayton** [comment]
We need to have a way to set the internal to either be "up first" or "down first".  So scaling old down by X percent first, then scaling new up by Y.  This is required for us to support scale in place when # replicas = available target nodes.

File: pkg/deploy/api/validation/validation.go [hunk 4]
```diff
@@ -180,6 +180,10 @@ func validateRollingParams(params *deployapi.RollingDeploymentStrategyParams) fi
 		errs = append(errs, fielderrors.NewFieldInvalid("timeoutSeconds", *params.TimeoutSeconds, "must be >0"))
 	}
 
+	if params.UpdatePercent != nil && (*params.UpdatePercent < 1 || *params.UpdatePercent > 100) {
```
Comment (kargakis) [inline pkg/deploy/api/validation/validation.go]: You are missing 1 and 100 here. So 
```go
(*params.UpdatePercent <= 1 || *params.UpdatePercent >= 100)
```

File: pkg/deploy/api/validation/validation_test.go [hunk 5]
```diff
@@ -365,6 +365,40 @@ func TestValidateDeploymentConfigMissingFields(t *testing.T) {
 			fielderrors.ValidationErrorTypeRequired,
 			"template.strategy.rollingParams.pre.failurePolicy",
 		},
+		"small template.strategy.rollingParams.updatePercent": {
```
Comment (kargakis) [inline pkg/deploy/api/validation/validation_test.go]: Shouldn't we have tests that pass somewhere? If you add those, test using the limits.

File: pkg/deploy/strategy/support/lifecycle.go [hunk 6]
```diff
@@ -262,11 +279,13 @@ func (c *FirstContainerReady) Accept(deployment *kapi.ReplicationController) err
...
 
-	// If we found multiple, use the first one and log a warning.
-	// TODO: should finding multiple be an error?
-	pod := &pods.Items[0]
-	if len(pods.Items) > 1 {
-		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
+	// Use the first unknown pod we find, and track all pods we see.
+	var pod *kapi.Pod
+	for _, candidate := range pods.Items {
+		if !c.knownPods.Has(candidate.Name) && pod == nil {
```
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: Nit but is there a reason you are following this circuit? If not, move the nil check in front (cheaper since we will avoid some map lookups).

File: pkg/deploy/strategy/rolling/rolling.go [hunk 7]
```diff
@@ -111,7 +111,7 @@ func NewRollingDeploymentStrategy(namespace string, client kclient.Interface, co
 			},
 		},
 		getUpdateAcceptor: func(timeout time.Duration) kubectl.UpdateAcceptor {
-			return stratsupport.NewFirstContainerReady(client, timeout, NewFirstContainerReadyInterval)
+			return stratsupport.NewOnePodReady(client, timeout, NewFirstContainerReadyInterval)
```
Comment (kargakis) [inline pkg/deploy/strategy/rolling/rolling.go]: s/NewFirstContainerReadyInterval/NewOnePodReadyInterval ?

File: pkg/deploy/strategy/rolling/rolling.go [hunk 8]
```diff
@@ -205,16 +205,22 @@ func (s *RollingDeploymentStrategy) Deploy(from *kapi.ReplicationController, to
...
+		UpdatePercent:  params.UpdatePercent,
 		CleanupPolicy:  kubectl.PreserveRollingUpdateCleanupPolicy,
 		UpdateAcceptor: updateAcceptor,
 	}
-	glog.Infof("Starting rolling update from %s to %s (desired replicas: %d, UpdatePeriodSeconds=%d, IntervalSeconds=%d, TimeoutSeconds=%d)",
+	pct := "<nil>"
+	if params.UpdatePercent != nil {
+		pct = fmt.Sprintf("%d", *params.UpdatePercent)
+	}
+	glog.Infof("Starting rolling update from %s to %s (desired replicas: %d, updatePeriodSeconds=%d, intervalSeconds=%d, timeoutSeconds=%d, updatePercent=%s)",
```
Comment (kargakis) [inline pkg/deploy/strategy/rolling/rolling.go]: Add a *%* in the end of updatePercent? Also *s* in the end of *Seconds would be nice.

**Comment by kargakis** [comment]
A validation fix and nits

File: pkg/deploy/api/validation/validation.go (hunk 4 above)
Comment (ironcladlou) [inline pkg/deploy/api/validation/validation.go]: But 1 and 100 are valid values...

File: pkg/deploy/strategy/support/lifecycle.go (hunk 6 above)
Comment (ironcladlou) [inline pkg/deploy/strategy/support/lifecycle.go]: Reversed the condition order.

File: pkg/deploy/api/validation/validation.go (hunk 4 above)
Comment (kargakis) [inline pkg/deploy/api/validation/validation.go]: Nvmd, it was early in the morning.

File: pkg/deploy/api/validation/validation_test.go (hunk 5 above)
Comment (ironcladlou) [inline pkg/deploy/api/validation/validation_test.go]: Refactored the tests slightly to support valid cases. Upper/lower bound validity and errors are now tested.

File: pkg/deploy/strategy/rolling/rolling.go (hunk 8 above)
Comment (ironcladlou) [inline pkg/deploy/strategy/rolling/rolling.go]: Done

File: pkg/deploy/strategy/rolling/rolling.go (hunk 7 above)
Comment (ironcladlou) [inline pkg/deploy/strategy/rolling/rolling.go]: OnePodReadyPerInterval?

**Comment by ironcladlou** [comment]
@smarterclayton 
>We need to have a way to set the internal to either be "up first" or "down first". So scaling old down by X percent first, then scaling new up by Y. This is required for us to support scale in place when # replicas = available target nodes.

This needs to be a followup issue.

File: pkg/deploy/strategy/rolling/rolling.go (hunk 7 above)
Comment (kargakis) [inline pkg/deploy/strategy/rolling/rolling.go]: sgtm

Comment (ironcladlou) [inline pkg/deploy/strategy/rolling/rolling.go]: Renamed

**Comment by ironcladlou** [comment]
[test]

**Comment by smarterclayton** [comment]
Why?

On Jul 15, 2015, at 9:00 AM, Dan Mace <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton>

We need to have a way to set the internal to either be "up first" or "down
first". So scaling old down by X percent first, then scaling new up by Y.
This is required for us to support scale in place when # replicas =
available target nodes.

This needs to be a followup issue.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-121607043>.

File: pkg/deploy/strategy/rolling/rolling.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/deploy/strategy/rolling/rolling.go]: It may not be enough to support one pod.  Imagine you have 100 pods and a
percent of 10.  The first pod completes (canary), then the second 10
start.  Only the first goes to ready.  The second 10 start.  Only the first
goes to ready, but by this time only 5 of the original 20 have started.  At
this point, only 85% of the pods are running.  Does a user want this?

The algorithm probably has to be that each chunk you grab has to be above
the percentage you specified as you change interval.  Ie, if you have 90
out of 100 ready and 10% as your interval, you need to wait.

On Jul 15, 2015, at 8:59 AM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/strategy/rolling/rolling.go
<https://github.com/openshift/origin/pull/3701#discussion_r34674822>:

> @@ -111,7 +111,7 @@ func NewRollingDeploymentStrategy(namespace string, client kclient.Interface, co
>  			},
>  		},
>  		getUpdateAcceptor: func(timeout time.Duration) kubectl.UpdateAcceptor {
> -			return stratsupport.NewFirstContainerReady(client, timeout, NewFirstContainerReadyInterval)
> +			return stratsupport.NewOnePodReady(client, timeout, NewFirstContainerReadyInterval)

OnePodReadyPerInterval?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701/files#r34674822>.

**Comment by ironcladlou** [comment]
@smarterclayton 
>Why?

Because that's a fundamental and orthogonal change to the upstream mechanism that we hadn't previously discussed.

The way we designed the current UpdateAcceptor patch we're carrying is "only accept once regardless of the interval count," then during the tunables work the requirements changed to "accept every interval," which led to me having to refactor the acceptor implementation as you see here. Now we're getting into even more unanticipated design discussion about the newly refactored update acceptor which I hadn't even intended to change for the tunables issue in the first place.

I'm already pretty uncomfortable rolling the UpdateAcceptor and tunables work together. I don't want to introduce a third change which really could be made separately after we've got some agreement on the preceding designs on which it would be based.

**Comment by smarterclayton** [comment]
Delivering features that don't quite work doesn't help us.  Can you tell me
what the design for the initial tunable I asked for would work?  The scale
down first vs scale up first is a blocker for registries and routers.
Percentages are just nice to haves.

On Wed, Jul 15, 2015 at 10:52 AM, Dan Mace <notifications@github.com> wrote:

> @smarterclayton <https://github.com/smarterclayton>
>
> Why?
>
> Because that's a fundamental and orthogonal change to the upstream
> mechanism that we hadn't previously discussed.
>
> The way we designed the current UpdateAcceptor patch we're carrying is
> "only accept once regardless of the interval count," then during the
> tunables work the requirements changed to "accept every interval," which
> led to me having to refactor the acceptor implementation as you see here.
> Now we're getting into even more unanticipated design discussion about the
> newly refactored update acceptor which I hadn't even intended to change for
> the tunables issue in the first place.
>
> I'm already pretty uncomfortable rolling the UpdateAcceptor and tunables
> work together. I don't want to introduce a third change which really could
> be made separately after we've got some agreement on the preceding designs
> on which it would be based.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3701#issuecomment-121641309>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ironcladlou** [comment]
Is there some prior reference to the scale down then up requirement? Sounds like we need a new issue and card to take priority over this one.

**Comment by smarterclayton** [comment]
I would just make -% alter the order (down before up)

On Wed, Jul 15, 2015 at 11:24 AM, Dan Mace <notifications@github.com> wrote:

> Is there some prior reference to the scale down then up requirement?
> Sounds like we need a new issue and card to take priority over this one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3701#issuecomment-121651011>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ironcladlou** [comment]
>I would just make -% alter the order (down before up)

The current upstream rolling updater always downs up then down. Are you suggesting I reverse that for all cases? Seems like that would warrant some discussion with other upstream folks involved with the rolling updater. Or did I misunderstand?

**Comment by smarterclayton** [comment]
I'm saying we have to support down then up.

On Jul 15, 2015, at 12:28 PM, Dan Mace <notifications@github.com> wrote:

I would just make -% alter the order (down before up)

The current upstream rolling updater always downs up then down. Are you
suggesting I reverse that for all cases? Seems like that would warrant some
discussion with other upstream folks involved with the rolling updater. Or
did I misunderstand?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-121668654>.

**Comment by smarterclayton** [comment]
As well as up then down.

On Jul 15, 2015, at 12:28 PM, Dan Mace <notifications@github.com> wrote:

I would just make -% alter the order (down before up)

The current upstream rolling updater always downs up then down. Are you
suggesting I reverse that for all cases? Seems like that would warrant some
discussion with other upstream folks involved with the rolling updater. Or
did I misunderstand?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-121668654>.

**Comment by ironcladlou** [comment]
Now that https://github.com/GoogleCloudPlatform/kubernetes/pull/10062 has an acceptable implementation for percentages and directionality, we should revisit @smarterclayton's `UpdateAcceptor` comment:

>It may not be enough to support one pod. Imagine you have 100 pods and a percent of 10. The first pod completes (canary), then the second 10 start. Only the first goes to ready. The second 10 start. Only the first goes to ready, but by this time only 5 of the original 20 have started. At this point, only 85% of the pods are running. Does a user want this? The algorithm probably has to be that each chunk you grab has to be above the percentage you specified as you change interval. Ie, if you have 90 out of 100 ready and 10% as your interval, you need to wait.

Should the update acceptor have its own independent percentage field on the rolling params, so that users can specify what percentage of each interval must be ready before proceeding (in addition to a timeout for the acceptance of the percentage)?

**Comment by smarterclayton** [comment]
My assumption is that the % is your input about your SLA.  If you're SLA is
90%, your update parameter should never be higher than 10%.  If that's
logical to end users, then the extra parameter doesn't help them understand
it better (I don't think).  Rather than have another percentage, what
you're really saying is "what defines when I should move to the next
chunk".  That's probably "when the pods are ready" or "when the pods are
up".  I don't know why you'd have a readiness check set if you didn't want
us to use it.

So I'd argue - treat 100 - update_percent as the SLA.  Don't go below that
(except due to fencepost / rounding on small numbers).

On Fri, Jul 17, 2015 at 5:09 PM, Dan Mace <notifications@github.com> wrote:

> Now that GoogleCloudPlatform/kubernetes#10062
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10062> has an
> acceptable implementation for percentages and directionality, we should
> revisit @smarterclayton <https://github.com/smarterclayton>'s
> UpdateAcceptor comment:
>
> It may not be enough to support one pod. Imagine you have 100 pods and a
> percent of 10. The first pod completes (canary), then the second 10 start.
> Only the first goes to ready. The second 10 start. Only the first goes to
> ready, but by this time only 5 of the original 20 have started. At this
> point, only 85% of the pods are running. Does a user want this? The
> algorithm probably has to be that each chunk you grab has to be above the
> percentage you specified as you change interval. Ie, if you have 90 out of
> 100 ready and 10% as your interval, you need to wait.
>
> Should the update acceptor have its own independent percentage field on
> the rolling params, so that users can specify what percentage of each
> interval must be ready before proceeding (in addition to a timeout for the
> acceptance of the percentage)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3701#issuecomment-122417355>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ironcladlou** [comment]
>My assumption is that the % is your input about your SLA. If you're SLA is 90%, your update parameter should never be higher than 10%. If that's logical to end users, then the extra parameter doesn't help them understand it better (I don't think). Rather than have another percentage, what you're really saying is "what defines when I should move to the next chunk". That's probably "when the pods are ready" or "when the pods are up". I don't know why you'd have a readiness check set if you didn't want us to use it. So I'd argue - treat 100 - update_percent as the SLA. Don't go below that (except due to fencepost / rounding on small numbers).

This makes sense to me. Doesn't that mean that the acceptance algo is "wait for all pods in the batch to become ready"?

**Comment by smarterclayton** [comment]
Yes

On Jul 20, 2015, at 10:36 AM, Dan Mace <notifications@github.com> wrote:

My assumption is that the % is your input about your SLA. If you're SLA is
90%, your update parameter should never be higher than 10%. If that's
logical to end users, then the extra parameter doesn't help them understand
it better (I don't think). Rather than have another percentage, what you're
really saying is "what defines when I should move to the next chunk".
That's probably "when the pods are ready" or "when the pods are up". I
don't know why you'd have a readiness check set if you didn't want us to
use it. So I'd argue - treat 100 - update_percent as the SLA. Don't go
below that (except due to fencepost / rounding on small numbers).

This makes sense to me. Doesn't that mean that the acceptance algo is "wait
for all pods in the batch to become ready"?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-122904308>.

**Comment by ironcladlou** [comment]
The `UpdateAcceptor` implementation has been totally redone so that it verifies *all new pods* it sees are ready before accepting. More details in [the comments](https://github.com/ironcladlou/origin/blob/rolling-updater-enhancements/pkg/deploy/strategy/support/lifecycle.go#L256).

**Comment by smarterclayton** [comment]
Thought experiment - for us to complete a deployment, 100-N% pods have to be ready.  I don't think we would fail a deployment (or block progress) if <N% are not ready (when update percentage is negative) or 100% (when update percentage is positive).  How hard is that to implement?  Basically we just need to maintain the SLA invariant (must always have that 100-N% running).  Can we do that safely?

File: pkg/deploy/strategy/support/lifecycle.go [hunk 9]
```diff
@@ -239,100 +239,107 @@ func NewFirstContainerReady(kclient kclient.Interface, timeout time.Duration, in
...
 		// If all containers have been ready once, return success.
-		if allReady {
-			glog.V(0).Infof("All containers ready for %s/%s", pod.Namespace, pod.Name)
+		if unreadyPods == 0 {
+			glog.V(0).Infof("Finished; %d new pods ready for %s", len(observedPodContainers), deployutil.LabelForDeployment(deployment))
 			return true, nil
 		}
 		// Otherwise, try again later.
-		glog.V(4).Infof("Still waiting for pod %s/%s container readiness; observed statuses: #%v", pod.Namespace, pod.Name, observedContainers)
+		glog.V(4).Infof("Still waiting for %d of %s pods to become ready for deployment %s", len(observedPodContainers), unreadyPods, unreadyContainers, deployutil.LabelForDeployment(deployment))
```
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: This log needs vetting

Comment (ironcladlou) [inline pkg/deploy/strategy/support/lifecycle.go]: I really wanted to emit progress at `V(0)` but I'm not sure of a great/all-purpose way to rate limit the output when the interval is short. Thoughts?

**Comment by ironcladlou** [comment]
>Thought experiment - for us to complete a deployment, 100-N% pods have to be ready. I don't think we would fail a deployment (or block progress) if <N% are not ready (when update percentage is negative) or 100% (when update percentage is positive). How hard is that to implement? Basically we just need to maintain the SLA invariant (must always have that 100-N% running). Can we do that safely?

So imagine this scenario:

```
UpdatePercent: -10%
SLA: 90%
Desired replicas: 100

i  old  new  ready(old+new)

0  100   0   100 
1  90   10   90  <-- accepted without waiting for readiness, SLA still met
2  80   20   89  <-- only 9 of i=1 are ready
```

Since the full percent increment will *always be scaled down to the SLA threshold* right before scale up/acceptance, I don't yet an alternative to waiting for all of the newly scaled-up pods to be ready. If we don't fully replace the scaled-down chunk from the start of the interval, we risk dipping below the SLA by our unready pods racing with the next interval's scale-down.

File: pkg/deploy/strategy/support/lifecycle.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/deploy/strategy/support/lifecycle.go]: We should spawn a follow up on the usability  of output.  Let's get this
closedz

On Jul 21, 2015, at 9:11 AM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/strategy/support/lifecycle.go
<https://github.com/openshift/origin/pull/3701#discussion_r35099349>:

>  			return true, nil
>  		}
>  		// Otherwise, try again later.
> -		glog.V(4).Infof("Still waiting for pod %s/%s container readiness; observed statuses: #%v", pod.Namespace, pod.Name, observedContainers)
> +		glog.V(4).Infof("Still waiting for %d of %s pods to become ready for deployment %s", len(observedPodContainers), unreadyPods, unreadyContainers, deployutil.LabelForDeployment(deployment))

I really wanted to emit progress at V(0) but I'm not sure of a
great/all-purpose way to rate limit the output when the interval is short.
Thoughts?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701/files#r35099349>.

**Comment by smarterclayton** [comment]
What if we altered the scale interval to account for unready?

I don't want to fail a 100 pod deploy because a node is bad.  Which is what
happens now.

On Jul 21, 2015, at 10:02 AM, Dan Mace <notifications@github.com> wrote:

Thought experiment - for us to complete a deployment, 100-N% pods have to
be ready. I don't think we would fail a deployment (or block progress) if
<N% are not ready (when update percentage is negative) or 100% (when update
percentage is positive). How hard is that to implement? Basically we just
need to maintain the SLA invariant (must always have that 100-N% running).
Can we do that safely?

So imagine this scenario:

UpdatePercent: -10%
SLA: 90%
Desired replicas: 100

i  old  new  ready(old+new)

0  100   0   100
1  90   10   90  <-- accepted without waiting for readiness, SLA still met
2  80   20   89  <-- only 9 of i=1 are ready

Since the full percent increment will *always be scaled down to the SLA
threshold* right before scale up/acceptance, I don't yet an alternative to
waiting for all of the newly scaled-up pods to be ready. If we don't fully
replace the scaled-down chunk from the start of the interval, we risk
dipping below the SLA by our unready pods racing with the next interval's
scale-down.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-123319319>.

**Comment by ironcladlou** [comment]
>What if we altered the scale interval to account for unready?

I don't follow... elaborate more on that idea?

>I don't want to fail a 100 pod deploy because a node is bad. Which is what happens now.

Yeah. Still having trouble imagining how to solve it. To deconstruct the "node death" case further: if the node dies but the pods can still be scheduled elsewhere, there's no problem (other than the time counted against the timeout). If the node death means we can't schedule some of the pods for an interval, what recourse do we have? We can't predict if a future interval will fill in the gap if we were to ignore it, so we have to assume that the current interval is our only chance to keep us above the line. Maybe whatever you had in mind for scale alteration above relates to this.

**Comment by smarterclayton** [comment]
On Jul 21, 2015, at 10:30 AM, Dan Mace <notifications@github.com> wrote:

What if we altered the scale interval to account for unready?

I don't follow... elaborate more on that idea?

I don't want to fail a 100 pod deploy because a node is bad. Which is what
happens now.

Yeah. Still having trouble imagining how to solve it. To deconstruct the
"node death" case further: if the node dies but the pods can still be
scheduled elsewhere, there's no problem (other than the time counted
against the timeout). If the node death means we can't schedule some of the
pods for an interval, what recourse do we have? We can't predict if a
future interval will fill in the gap if we were to ignore it, so we have to
assume that the current interval is our only chance to keep us above the
line. Maybe whatever you had in mind for scale alteration above relates to
this.


Treat update percent as the max scale interval.  Only scale down to the
SLA. If zero can be scaled down, that's when the timeout window starts
(timeout defined as the last interval we could move forward on).

So if you have +10%, scale up, wait until pods are ready, scaled down
min(10% * N, ready - N).  If you have -10%, scale down, scale up, as soon
as you have any ready do the next scale down, then scale up.  By the time
you get to the second round, more pods from the first round have become
ready.

Basically, don't scale by fixed chunks.  Scaled based on how many are ready.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-123338362>.

**Comment by ironcladlou** [comment]
>Treat update percent as the max scale interval. Only scale down to the SLA. If zero can be scaled down, that's when the timeout window starts (timeout defined as the last interval we could move forward on). So if you have +10%, scale up, wait until pods are ready, scaled down min(10% * N, ready - N). If you have -10%, scale down, scale up, as soon as you have any ready do the next scale down, then scale up. By the time you get to the second round, more pods from the first round have become ready. Basically, don't scale by fixed chunks. Scaled based on how many are ready.

I'm pretty sure this is all clear to me now and seems workable. Unfortunately, now we're back to...

> How hard is that to implement?

The upstream code would need re-designed with this in mind. It would need to take on the responsibility of checking overall readiness each interval to recompute the scaling increment, or rely on the update acceptor to give it new values (or context to inform the computation), or something else.

**Comment by smarterclayton** [comment]
Let's finish this out and treat at as a next step.  What we have here is
strictly all necessary - we have introduced the concept that we don't
violate an SLA.  But we are now in the case where one failure breaks the
CLA.

On Jul 21, 2015, at 10:56 AM, Dan Mace <notifications@github.com> wrote:

Treat update percent as the max scale interval. Only scale down to the SLA.
If zero can be scaled down, that's when the timeout window starts (timeout
defined as the last interval we could move forward on). So if you have
+10%, scale up, wait until pods are ready, scaled down min(10% * N, ready -
N). If you have -10%, scale down, scale up, as soon as you have any ready
do the next scale down, then scale up. By the time you get to the second
round, more pods from the first round have become ready. Basically, don't
scale by fixed chunks. Scaled based on how many are ready.

I'm pretty sure this is all clear to me now and seems workable.
Unfortunately, now we're back to...

How hard is that to implement?

The upstream code would need re-designed with this in mind. It would need
to take on the responsibility of checking overall readiness each interval
to recompute the scaling increment, or rely on the update acceptor to give
it new values (or context to inform the computation), or something else.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-123358633>.

**Comment by ironcladlou** [comment]
@kargakis

I drastically simplified the acceptor implementation based on your suggestion to use `kapi.IsPodReady`. PTAL.

**Comment by smarterclayton** [comment]
"One failure breaks the deploy" which is incredibly fragile for any deployment over 5 or 10 pods.  Follow?  Probability of independent failures grows independently.

File: pkg/deploy/strategy/support/lifecycle.go [hunk 10]
```diff
@@ -239,100 +239,75 @@ func NewFirstContainerReady(kclient kclient.Interface, timeout time.Duration, in
...
-				observedContainers[status.Name] = true
-			}
-		}
-		// Check whether all containers have been observed as ready.
-		allReady := true
-		for _, ready := range observedContainers {
-			if !ready {
-				allReady = false
-				break
+			if kapi.IsPodReady(pod) {
```
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: This will check all pods every time right? Do we want that or we just want to check a pod once and then never check it again?

Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: If the latter, make this
```go
if c.acceptedPods.Has(pod) && kapi.IsPodReady(pod) {
```

Comment (smarterclayton) [inline pkg/deploy/strategy/support/lifecycle.go]: I think the latter.  But you can't scale down if one of your new ones goes
down without breaching SLA.

On Jul 21, 2015, at 12:09 PM, Michail Kargakis <notifications@github.com>
wrote:

In pkg/deploy/strategy/support/lifecycle.go
<https://github.com/openshift/origin/pull/3701#discussion_r35120393>:

>  				continue
>  			}
> -			// The status of the container could be transient; we only care if it
> -			// was ever ready. If it was ready and then became not ready, we
> -			// consider it ready.
> -			if status.Ready {
> -				observedContainers[status.Name] = true
> -			}
> -		}
> -		// Check whether all containers have been observed as ready.
> -		allReady := true
> -		for _, ready := range observedContainers {
> -			if !ready {
> -				allReady = false
> -				break
> +			if kapi.IsPodReady(pod) {

If the latter, make this

if c.acceptedPods.Has(pod) && kapi.IsPodReady(pod) {

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701/files#r35120393>.

Comment (ironcladlou) [inline pkg/deploy/strategy/support/lifecycle.go]: It's not equivalent. Still have to account for if the pod wasn't in accepted pods and is unready to insert it into the unready set. I don't think that would be any more clear or less code. It's already compact.

The current code doesn't perfom a readiness check on previously accepted pods.

I think the logic I have is correct.

**Comment by ironcladlou** [comment]
>"One failure breaks the deploy" which is incredibly fragile for any deployment over 5 or 10 pods. Follow? Probability of independent failures grows independently.

Yes.

**Comment by smarterclayton** [comment]
You and I can chat on that after we land this.  I think this is LGTM but
need to take one more pass

On Jul 21, 2015, at 1:16 PM, Dan Mace <notifications@github.com> wrote:

"One failure breaks the deploy" which is incredibly fragile for any
deployment over 5 or 10 pods. Follow? Probability of independent failures
grows independently.

Yes.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701#issuecomment-123406555>.

File: pkg/deploy/strategy/support/lifecycle.go (hunk 10 above)
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: Oh I missed your check above

File: pkg/deploy/strategy/support/lifecycle.go [hunk 11]
```diff
@@ -239,100 +239,75 @@ func NewFirstContainerReady(kclient kclient.Interface, timeout time.Duration, in
...
-	}
-
-	// If we found multiple, use the first one and log a warning.
-	// TODO: should finding multiple be an error?
-	pod := &pods.Items[0]
-	if len(pods.Items) > 1 {
-		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
-	}
-
+func (c *AcceptNewlyObservedReadyPods) Accept(deployment *kapi.ReplicationController) error {
```
Comment (kargakis) [inline pkg/deploy/strategy/support/lifecycle.go]: godoc?

Comment (ironcladlou) [inline pkg/deploy/strategy/support/lifecycle.go]: done ;)

**Comment by smarterclayton** [comment]
This is LGTM [merge]

**Comment by smarterclayton** [comment]
Link the next issue here please

On Wed, Jul 22, 2015 at 1:44 PM, OpenShift Bot <notifications@github.com>
wrote:

> Merged #3701 <https://github.com/openshift/origin/pull/3701>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3701#event-362550531>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ironcladlou** [comment]
Created #3844 as a followup.
